File: T2/T2_EjemploPersonas/app/src/main/java/com/system/develop/t2_ejemplopersonas/adaptadores/AdaptadorPersonalizado.java
package com.system.develop.t2_ejemplopersonas.adaptadores;

import android.app.Person;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.system.develop.t2_ejemplopersonas.R;
import com.system.develop.t2_ejemplopersonas.utils.Persona;

import java.util.List;

public class AdaptadorPersonalizado extends BaseAdapter {

    List<Persona> listaDatos;
    Context context;

    public AdaptadorPersonalizado(List<Persona> listaDatos, Context context) {
        this.listaDatos = listaDatos;
        this.context = context;
    }

    @Override
    public int getCount() {
        return listaDatos.size();
    }

    @Override
    public Object getItem(int position) {
        return listaDatos.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    public void agregarPersona(Persona persona){
        listaDatos.add(persona);
        notifyDataSetChanged();
    }

    public void borrarPersona(int posicion){
        listaDatos.remove(posicion);
        notifyDataSetChanged();
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {

        HolderPropio holder;

        if(convertView==null){
            convertView = LayoutInflater.from(context).inflate(R.layout.item_lista,parent,false);
            holder = new HolderPropio(convertView);
            convertView.setTag(holder);
        }
        else {
           holder = (HolderPropio) convertView.getTag();
        }



        Persona p = listaDatos.get(position);
        holder.getNombreHolder().setText(p.getNombre());
        holder.getApellidoHolder().setText(p.getApellido());
        holder.getNombreHolder().setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                borrarPersona(position);
                return true;
            }
        });

        return convertView;
    }

    class HolderPropio{
        TextView nombreHolder;
        TextView apellidoHolder;

        public HolderPropio(View view) {
            nombreHolder = view.findViewById(R.id.nombrePersona);
            apellidoHolder = view.findViewById(R.id.apellidoPersona);
        }

        public TextView getNombreHolder() {
            return nombreHolder;
        }

        public TextView getApellidoHolder() {
            return apellidoHolder;
        }
    }
}

File: Kotlin/T3/codigos2223/T3_Spinner/app/src/main/java/com/develop/t3_spinner/adaptadores/AdaptadorPersonalizado.kt
package com.develop.t3_spinner.adaptadores

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import com.develop.t3_spinner.R
import com.develop.t3_spinner.modelo.Pais

class AdaptadorPersonalizado(var listaElementos: ArrayList<Pais>, var contexto: Context) : BaseAdapter() {


    override fun getCount(): Int {
        // retornar el número de elementos que el spinner tiene que pintar
        return listaElementos.size;
    }

    override fun getItem(p0: Int): Any {
        // retorna el elemento que esta en una posicion indicada como parametro
        return listaElementos.get(p0)
    }

    override fun getItemId(p0: Int): Long {
        // retorna el long asociado a un elemento que esta en una posicion indicada como parametro
        return p0.toLong();
    }

    fun agregarPais(pais: Pais){
        listaElementos.add(pais)
        notifyDataSetChanged();
    }

    override fun getView(p0: Int, p1: View?, p2: ViewGroup?): View {
        // retona la vista (fila) ya rellena con los datos que estas
        // en una posición indicada -> p0
        // para poder rellenar la fila -> XML

        var vista: View =  LayoutInflater.from(contexto).inflate(R.layout.item_spinner,p2,false)

        var nombre: TextView = vista.findViewById(R.id.texto_fila);
        var imagen: ImageView = vista.findViewById(R.id.imagen_fila);

        var elemento = listaElementos.get(p0)
        nombre.text = elemento.nombre
        imagen.setImageResource(elemento.imagen)

        return vista;

    }
}